server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -680,72 +689,181 @@ public String toString() {
+         *
+         * This implementation can send up to {@code maxConcurrentFileChunks} consecutive file-chunk requests without waiting
+         * for the replies from the recovery target to reduce the recovery time in secure/compressed/high latency communication.
+         * We assign a seqId to every file-chunk request and stop reading/sending file chunks if the gap between max_seq_no
+         * and local_checkpoint is greater than {@code maxConcurrentFileChunks}.
+         */
+        void sendFileChunks(ActionListener<Void> listener) {
+            while (true) {
+                cancellableThreads.checkForCancel();
+                synchronized (this) {
+                    if (semaphore.tryAcquire() == false) {
                         break;
                     }
-                    final long requestFilePosition = position;
-                    cancellableThreads.executeIO(() ->
-                        recoveryTarget.writeFileChunk(md, requestFilePosition, content, lastChunk, translogOps.get(),
+                    // don't send more, the number of unreplied chunks is already greater than maxConcurrentFileChunks,
+                    if (requestSeqIdTracker.getMaxSeqNo() - requestSeqIdTracker.getCheckpoint() >= maxConcurrentFileChunks) {
+                        semaphore.release();
+                        break;
+                    }
+                }
+                try {
+                    final byte[] reusedBuffer = recycledBuffers.pollFirst();
+                    final byte[] buffer = reusedBuffer != null ? reusedBuffer : new byte[chunkSizeInBytes];
+                    final FileChunk chunk = readChunk(buffer);
+                    semaphore.release(); // other thread can read and send chunks
+                    if (chunk == null) {
+                        if (requestSeqIdTracker.getMaxSeqNo() == requestSeqIdTracker.getCheckpoint()) {
+                            listener.onResponse(null);
+                        }
+                        break;
+                    }
+                    cancellableThreads.execute(() ->
+                        recoveryTarget.writeFileChunk(chunk.md, chunk.position, chunk.content, chunk.lastChunk, translogOps.get(),
                             ActionListener.wrap(
-                                r -> requestSeqIdTracker.markSeqNoAsCompleted(requestSeqId),
-                                e -> {
-                                    error.compareAndSet(null, Tuple.tuple(md, e));
-                                    requestSeqIdTracker.markSeqNoAsCompleted(requestSeqId);
-                                }
+                                r -> {
+                                    recycledBuffers.addFirst(buffer);
+                                    requestSeqIdTracker.markSeqNoAsCompleted(chunk.seqId);
+                                    sendFileChunks(listener);
```

Comment:
can we make sure we always fork here somehow? I am a bit worried that we are ending up with a stack overflow? Like we can assert that we don't have `sendFileChunks` in the stacktrace for instance.

Reply:
I opened https://github.com/elastic/elasticsearch/pull/39988 for this.